### ISSUE TITLE
High cold star times

### ISSUE THREAD

**Issue by ari-luokkala** [serverless-nextjs/serverless-next.js#1997]
Describe the Bug

With the use of experimental-serverless-trace with serverless-next.js the SSR page Lambda@Edge cold boot times are roughly duplicated. This causes application performance downgrade on lambda cold starts serving that page.

Expected Behavior

Boot times are roughly the same with serverless and experimental-serverless-trace targets.

To Reproduce

Set useServerlessTraceTarget to true in serveless.yml.

**Comment by dphang** [comment]
Please follow the bug template in the future and include the versions of Next.js and this component's versions you are using.

Yes, in older versions of Next.js (I think before 11) it has high cold start time because the `node_modules` were being traced and not bundled/tree-shaken. AFAIK in newer versions Next.js uses new shared chunks of JS even for serverless trace target so this should not be happening now

**Comment by jonahallibone** [comment]
@dphang what is the default target used by this package when you deploy? I was under the impression (I think now wrongly) that it was the old serverless target, with an option for experiential-serverless-trace.

**Comment by dphang** [comment]
We are using the `serverless` target by default, which as of new Next.js versions (I think 11+) is using shared SSR chunks, and it seems to be the same as `experimental-serverless-trace` now, so not sure if it's doing much?

As mentioned, my problem with the previous behavior of serverless-trace target was that we needed to use `@vercel/nft` to trace the dependencies once (e.g node modules), although it was not tree-shaking things so for example packages like the entire `@material-ui` may get included and slowed perf a lot due to cold starts.

Not sure what the plan is from Vercel but it seems like the `serverless-trace` target is going away?

**Comment by jonahallibone** [comment]
@dphang honestly the documentation I've seen from Vercel on this front has been incredibly confusing. Sometimes it seems like `serverless` is going away and `serverless-trace` is [staying](https://github.com/vercel/next.js/issues/20487). In that comment Tim cites this repo as a reason for keeping `serverless`, although at this point it seems like it's just not the right default option anymore? I've probably read every thread like 10 times and still don't quite understand what their approach is.   As you've pointed out before, it seems like `serverless-trace` is being removed as well though, from comments by Tim. It does seem like Next 12+ does include some new tracing, not sure if you've noticed any differences

**Comment by ari-luokkala** [comment]
@dphang I refreshed the description to contain the correct bug template now.

**Comment by jonahallibone** [comment]
@dphang not to blow up this thread, but I found a new new useful things:
1. `target` itself is going away, not just the `serverless` or `experimental-serverless-trace` [here](https://nextjs.org/blog/next-12#output-file-tracing)
2. There is now documentation on how to use `@vercel/nft` and seems like it's what they recommend using for serverless environments. [here](https://nextjs.org/docs/advanced-features/output-file-tracing)

**Comment by dphang** [comment]
@ari-luokkala thanks. Any particular dependencies you are using and can you share a simple repo with? Wondering if experimental-serverless-trace is not efficiently bundling now...

@jonahallibone thanks for the update, I hadn't read into that latest update yet. I will look into that more. Actually that could be beneficial if it's now more efficiently tree shaking now - last time there was poor performance because it `@vercel/nft` was not properly tree shaking the MaterialUI library

**Comment by jonahallibone** [comment]
@dphang they also added two exportable functions where you can ignore/add dependencies which might not be properly bundled

**Comment by ari-luokkala** [comment]
@dphang I can provide some specific files and configs, but I need to know what exact files would suffice.

**Comment by dphang** [comment]
@ari-luokkala Yes, I think it would be good to know which node dependencies (if any) you are using that could cause the higher cold starts...otherwise I can try to repro later myself but I hadn't used this target myself for a while so would take some time. It would be good if you can try to compare the generated code first and get a sense of what the differences are?

I remember before MaterialUI had a problem with old serverless-trace, not sure if it's a similar thing.

**Comment by ari-luokkala** [comment]
@dphang I noticed that with trace target also `node_modules` are put to the deployment zip file, other than that files look the same, but did not check e.g. require blocks. However seems to be that this problem has disappeared in our environments and we need to investigate more. I will probably keep this issue open for a while, but will close as soon as we are sure that problem is gone.

**Comment by ari-luokkala** [comment]
@dphang It seems to be that the problem appears only when we need to do a NodeJS require('path/to/nextjs/page')  to render one of our pages server-side. With SSG this problem seems to be gone.

**Comment by dphang** [comment]
@ari-luokkala sorry I hadn't gotten a chance to look again, yes if `node_modules` I recall it's not bundled efficiently...I also had the same problem with material UI as mentioned, it seemed @vercel/nft does not tree-shake `node_modules` efficiently compared to webpack? You should check if you are using any large libraries where you might be inadvertently including code that's not actually used but adds to `require` time

**Comment by federicobarera** [comment]
How are you guys tracing the performance of the lambda@edge function (X-Ray doesn't seem applicable to @edge). 
We are facing performance issues that see ssr pages load time doubling when deployed @edge (compared to ec2/ecs deployments, same vcores and mem allocations).

We do have experimental-server-trace enabled due to the size limitations. However the performance degradation is consistent, hence I am trying to drill down were the bottleneck lies (I cannot pinpoint it to cold starts yet).

Suggestions?

**Comment by ari-luokkala** [comment]
We have used CloudFront monitoring and from there Lambda@Edge metrics. Additionally for debugging we some times fork this repo and deploy custom version with added custom debug logging. We have also set "logLambdaExecutionTimes: true" in serverless.yml, but I haven not noticed any impact with that. In the lambda logs, I think there is also this "initial time" or similar showing the cold boot times.

**Comment by khg0712** [comment]
Any updates here?
